sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -305,17 +306,19 @@ object ShuffleExchangeExec {
         rdd
       }
 
+      // round-robin function is order sensitive if we don't sort the input.
+      val orderSensitiveFunc = isRoundRobin && !SQLConf.get.sortBeforeRepartition
       if (needToCopyObjectsBeforeShuffle(part)) {
-        newRdd.mapPartitionsInternal { iter =>
+        newRdd.mapPartitionsWithIndexInternal((_, iter) => {
```

Comment:
Shouldn't we mark `newRdd` as `IDEMPOTENT` if insert a local sort (or `INDETERMINATE` if don't sort), so we don't have to mark the function as order sensitive?

Reply:
I agree it's clearer, the problem is we need to build a framework to set the property of map functions, e.g.
1. FORCE_ORDER: even the input order changed, the output data set and order will not change. (e.g. sort)
2. SAME_SET: even the input order changed, the output data set will not change. (e.g. i => i + 1)
3. ORDER_SENSITIVE: if the input order change, the output data set will be different. (e.g. round robin)

There may be more types, I haven't dug into it yet. Since it's only used here, maybe not worth to do it now?

Reply:
sounds good